File: app/src/main/java/xtr/keymapper/editor/EditorUiComponentList.java
package xtr.keymapper.editor;

import android.content.Context;

import androidx.annotation.MenuRes;

import java.util.ArrayList;
import java.util.function.Consumer;

import xtr.keymapper.R;
import xtr.keymapper.dpad.Dpad;
import xtr.keymapper.editor.component.Crosshair;
import xtr.keymapper.editor.component.Key;
import xtr.keymapper.editor.component.RightClick;
import xtr.keymapper.keymap.KeymapProfileKey;
import xtr.keymapper.swipekey.SwipeKey;

class EditorUiComponentList extends ArrayList<EditorUiComponent> {

    public Factory newFactory(EditorUiComponentCallback callback, Context context) {
        return new Factory(callback, context);
    }

    public void addMatchingComponentForId(@MenuRes int id, EditorUiComponentCallback callback, Context context, float x, float y) {
        final EditorUiComponent component;
        if (id == R.id.add) {
            component = new Key(callback, context, x, y);
        }
        else if (id == R.id.dpad) {
            component = new xtr.keymapper.editor.component.Dpad(callback, context, x, y).pickType();
        }
        else if (id == R.id.crosshair) {
            component = new Crosshair(callback, context, x, y);
        }
        else if (id == R.id.swipe_key) {
            component = new xtr.keymapper.editor.component.SwipeKey(callback, context, x, y);
        }
        else if (id == R.id.mouse_right) {
            component = new RightClick(callback, context, x, y);
        }
        else if (id == R.id.mouse_left) {
            component = null;
            forEachComponentOfClass(Crosshair.class,
                    crosshair -> crosshair.animateLeftClick(x, y));
        }
        else if (id == R.id.camera) {
            component = null;
        }
        else {
            component = null;
        }
        if (component != null) add(component);
    }

    public <T> void forEachComponentOfClass(Class<T> clazz, Consumer<? super T> action) {
        for (EditorUiComponent component : this) {
            if (clazz.isInstance(component)) {
                action.accept(clazz.cast(component));
            }
        }
    }



    class Factory {
        private final EditorUiComponentCallback mCallback;
        private final Context context;

        public Factory(EditorUiComponentCallback callback, Context context) {
            this.mCallback = callback;
            this.context = context;
        }

        public void addKeys(ArrayList<KeymapProfileKey> keys) {
            keys.forEach( key -> add(new Key(mCallback, context, key)));
        }

        public void addSwipeKeys(ArrayList<SwipeKey> swipeKeys) {
            swipeKeys.forEach(swipeKey -> add(new xtr.keymapper.editor.component.SwipeKey(mCallback, context, swipeKey)));
        }

        public void addDpads(Dpad[] dpadArray) {
            for (Dpad dpad: dpadArray) {
                if (dpad != null)
                    add(new xtr.keymapper.editor.component.Dpad(mCallback, context, dpad));
            }
        }

    }

}

File: app/src/main/java/xtr/keymapper/editor/EditorUiComponentList.kt
package xtr.keymapper.editor

// SoufianoDev: Standard Editor Imports And Component Dependencies.
import android.content.Context
import androidx.annotation.MenuRes
import xtr.keymapper.R
import xtr.keymapper.editor.component.*
import java.util.function.Consumer

internal class EditorUiComponentList : ArrayList<EditorUiComponent>() {

    // SoufianoDev: Initialize A New Component Factory For Shared Callback Context.
    fun newFactory(callback: EditorUiComponentCallback, context: Context): Factory {
        return Factory(callback, context)
    }

    /* Note : This Method Maps Menu IDs To Functional UI Components.
       The New Implementation Replaces Unsafe Double Bangs (!!) With Safe Calls
       And Enforces Non-Nullable Contexts To Enhance Editor Stability.
    */
    fun addMatchingComponentForId(
        @MenuRes id: Int,
        callback: EditorUiComponentCallback,
        context: Context,
        x: Float,
        y: Float,
        addComponent: (EditorUiComponent) -> Boolean
    ) = when (id) {
        R.id.add -> Key(callback, context, x, y)
        R.id.crosshair -> Crosshair(callback, context, x, y)
        R.id.swipe_key -> SwipeKey(callback, context, x, y)
        R.id.mouse_right -> RightClick(callback, context, x, y)
        R.id.camera -> Camera(callback, context, x, y)
        R.id.mouse_walk -> MouseWalk(callback, context, x, y)
        R.id.dpad -> {
            // SoufianoDev: Dpad Selection Requires A Context Picker Before Adding To Screen.
            Dpad(callback, context, x, y).pickType { addComponent(it) }
            null
        }
        R.id.mouse_left -> {
            // SoufianoDev: Animate Click Visuals Across All Active Crosshair Instances.
            forEachComponentOfClass(Crosshair::class.java) { crosshair ->
                crosshair?.animateLeftClick(x, y)
            }
            null
        }
        else -> null
    }?.let(addComponent)

    fun <T> forEachComponentOfClass(clazz: Class<T>, action: Consumer<in T?>) {
        for (component in this) {
            if (clazz.isInstance(component)) {
                action.accept(clazz.cast(component))
            }
        }
    }

    internal inner class Factory(
        private val mCallback: EditorUiComponentCallback,
        private val context: Context
    ) {
        // SoufianoDev: Batch Load Physical Keys Into The UI Editor Space.
        fun addKeys(keys: ArrayList<xtr.keymapper.keymap.element.Key?>) {
            keys.forEach { key ->
                if (key != null) add(Key(mCallback, context, key))
            }
        }

        // SoufianoDev: Batch Load Swipe Gestures Into The UI Editor Space.
        fun addSwipeKeys(swipeKeys: ArrayList<xtr.keymapper.keymap.element.SwipeKey?>) {
            swipeKeys.forEach { swipeKey ->
                if (swipeKey != null) add(SwipeKey(mCallback, context, swipeKey))
            }
        }

        fun addDpads(dpadArray: Array<xtr.keymapper.keymap.element.Dpad?>) {
            for (dpad in dpadArray) {
                if (dpad != null) add(Dpad(mCallback, context, dpad))
            }
        }
    }
}
